### PR TITLE
Introduce read-only panes

### LINKED ISSUE #6981: Read-only session mode to prevent accidentally killing a long-lived process
# Description of the new feature/enhancement

I sometimes run `ROBOCOPY` or another long-lived process in Windows Terminal. I then fear that I might accidentally close the tab or hit Ctrl+C and have to start over. I would like to mark the session or tab as protected (important and read-only) and have Windows Terminal protect me from such mistakes.

I already mark such tabs with a distinctive color but they still feel risky.

# Proposed technical implementation details (optional)

Add a "Read only" check box to the per-tab menu. When the check box is checked:

- Do not forward any keyboard events to processes using the pseudoconsole. Instead, if the user types something, display a modeless notification that advises the user how to disable the read-only mode.
- Do not forward any mouse events to processes using the pseudoconsole. Instead, let mouse events scroll the scrollback and select text. This should override the ENABLE_MOUSE_INPUT mode and perhaps also ENABLE_QUICK_EDIT_MODE.
- Do not display a close button on the tab. This may also make it consume less horizontal space, depending on the "tabWidthMode" setting.
- If the user tries to close the window, display a confirmation dialog box that includes the titles of the read-only tabs. This should override the "confirmCloseAllTabs" setting.
- If the user tries to close the tab (with a key combination, as the close button is hidden), then either display a notification as with keyboard events, or prompt for confirmation.

There could be a key-bindable action to turn the read-only mode on and off.

### PR DESCRIPTION
**Pull request by Don-Vito**
## Summary of the Pull Request
Introduces read-only panes.
When pane is marked as read-only:
1. Attempt to provide user input results in a warning
2. Attempt to close pane - shows dialog
3. Attempt to close hosting tab shows dialog
4. The hosting tab has no close button

## PR Checklist
* [x] Closes #6981
* [x] CLA signed. 
* [ ] Tests added/passed
* [ ] Documentation updated - not yet.
* [x] Schema updated.
* [ ] I've discussed this with core contributors already. 

## Detailed Description of the Pull Request / Additional comments
1. The readonly  logic implemented in `TermControl`
(and prevents any send input)
2. Special handling is required to allow key-bindings
3. The "close-readonly" protections are in TerminalPage.
4. The indication that the pane is readonly is done using lock glyph
5. The indication that the tab contains readonly pane
is done by hiding the close button of the tab
6. The readonly mode is enabled by keyboard shortcut
(the followup might add this to the context menu)

### CODE REVIEW

**Comment by Don-Vito** [comment]
ReadOnlyPanes2

Some highlights:
image

image

image

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:54]:
```suggestion
static constexpr std::string_view TogglePaneReadOnlyKey{ "toggleReadOnlyMode" };
```
Maybe? Curious what @cinnamon-msft thinks.

File: src/cascadia/TerminalControl/TermControl.h
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.h:270]:
```suggestion
        bool _isReadOnly;
```
nit: we generally initialize the value in the constructor.

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalTab.cpp:1126]:
```suggestion
```
Is there a follow-up to add a lock icon to the tab?
I'd remove this statement since that'll change soon, and wouldn't want to forget to update it. Teeny nit though.

File: src/cascadia/TerminalApp/Pane.h
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Pane.h:84]:
```suggestion
    bool ContainsReadOnly() const;
```

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Resources/en-US/Resources.resw:417]:
```suggestion
    <value>You are about to close a read-only terminal. Do you wish to continue?</value>
```
Maybe?

**Review by carlos-zamora — APPROVED** [review]
Some other notes:
- [ ] mildly concerned if "Close anyway" should be the primary button
- [ ] mildly concerned about some of the async changes to closing tabs
So, honestly, I'd want a design signoff from @cinnamon-msft and a closer look from @zadjii-msft (or @DHowett) for async tab closing.

Mostly nits, and I don't see any real issues, so I'm approving.

**Comment by Don-Vito** [comment]
> 
> 
> Some other notes:
> 
>     * [ ]  mildly concerned if "Close anyway" should be the primary button
> 
>     * [ ]  mildly concerned about some of the async changes to closing tabs
>       So, honestly, I'd want a design signoff from @cinnamon-msft and a closer look from @zadjii-msft (or @DHowett) for async tab closing.
> 
> 
> Mostly nits, and I don't see any real issues, so I'm approving.

Yeah. I am mostly worried about the  async nature of remove tab. Hope I didn't miss anything.
Regarding the primary button - I absolutely agree and will change it!

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (Don-Vito) [reply src/cascadia/TerminalApp/TerminalTab.cpp:1126]: So regarding this I am not sure if there will be a follow up - the lock icon appears only when the locked **pane** is focused.
But when the locked pane is not focused we want to somehow indicate that another pane in this tab is locked.

So right now we simply hide the close button. Alternatively, we could also add another glyph (instead of the lock) to indicate that inactive pane is locked.

WDYT?

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalTab.cpp:1126]: >  the lock icon appears only when the locked pane is focused... So right now we simply hide the close button

I think I'm okay with that UX personally. There's not usually _that many_ panes in a tab that tracking down the one that's read only would be that challenging...

File: src/cascadia/TerminalControl/TermControl.h
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/TermControl.h:270]: That's not.. strictly true. We've been back and forth on that. I think we've been moving towards constructing in the header when possible.

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:54]: I definitely prefer `toggleReadOnlyMode`

File: src/cascadia/TerminalControl/TermControl.h
Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControl.h:270]: _derp_, ignore me then haha

File: src/cascadia/TerminalControl/TermControl.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.cpp:894]: Should we add this check to the `F7` handler below too? Maybe just the whole method?

File: src/cascadia/TerminalControl/TermControl.idl
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.idl:123]: nit: should this be a typed handler like `event Windows.Foundation.TypedEventHandler<Object, Object> ReadOnlyChanged`? I feel like most of our recent events have been like that

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1199]: This comment feels like it should be down above the call to `Shutdown`, L1224 now

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1204]: Wait the `Primary` button is the "Cancel" button? That's... unexpected? confusing?

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1601]: Should we put this logic inside of `Tab::ClosePane`, to try and mitigate the chances for a race here?

I'm imagining in the post #4472 world, someone could totally `wt -w 1 split-pane` to create a new pane in _this_ window, and this race would be easy to repro

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1644]: Just as a sanity check, if we have 4 tabs, and the second tab is read-only, we'll
1. close tab 1 immediately
2. prompt for confirmation to close tab 2. (we may or may not end up closing it)
3. close tab 3 (after we figure out what we're doing with tab 2)
4. close tab 4 immediately after closing 3

Right?

**Review by zadjii-msft — COMMENTED** [review]
This mostly seems fine to me. I'm a little worried about the race in closing a pane. Hopefully if we move that logic within the tab itself, we can mitigate that.

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (Don-Vito) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1204]: How do you suggest to address this? Will comment be enough? Carlos and I agree that the safer "default choice" should be Cancel.

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1204]: Oh yea that makes sense. I guess I was just confused looking at the pics. Probably leaving a comment is more than enough to make sense of this ☺️

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (Don-Vito) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1601]: I guess we should. I was wondering the ROI :blush:

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (Don-Vito) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1644]: The tabs are closed one by one because we await each one of them

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (Don-Vito) [reply src/cascadia/TerminalApp/TerminalPage.cpp]: Done

**Comment by Don-Vito** [comment]
@zadjii-msft, @carlos-zamora - please re-review. Few additional thoughts:
1. I guess we need a tutorial for this. WDYT? If so - is it done by PM?
2. WDYT of providing a default key binding? If so - what should it be?

**Comment by carlos-zamora** [comment]
Tagging @cinnamon-msft since she's the PM

> 1. I guess we need a tutorial for this. WDYT? If so - is it done by PM?

This could go in the tips-and-tricks portion of the docs site. I don't feel strongly about that though. Being more descriptive in the 'actions' docs page should be sufficient, I think,

> 2. WDYT of providing a default key binding? If so - what should it be?

I don't think a default keybinding makes sense here. _However_, we should definitely add it to the command palette like this:
```js
{ "command": "toggleReadOnlyMode" }
```

**Comment by Don-Vito** [comment]
> Tagging @cinnamon-msft since she's the PM
> 
> > 1. I guess we need a tutorial for this. WDYT? If so - is it done by PM?
> 
> This could go in the tips-and-tricks portion of the docs site. I don't feel strongly about that though. Being more descriptive in the 'actions' docs page should be sufficient, I think,
> 
> > 2. WDYT of providing a default key binding? If so - what should it be?
> 
> I don't think a default keybinding makes sense here. _However_, we should definitely add it to the command palette like this:
> ```js
> { "command": "toggleReadOnlyMode" }
> ```
> 

Already added it to command palette

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalTab.cpp:1128]:
```suggestion
        const auto control = GetActiveTerminalControl();
```
nit

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1589]:
```suggestion
```

**Review by carlos-zamora — APPROVED** [review]
Still looks good to me.

**Comment by zadjii-msft** [comment]
(fyi I've got some things I need to address over the next 36 hours, but this is blocked off on my calendar for Thursday morning. Sorry for the delay 😬 )

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1612]: Yea, this was fine. Nothing to worry about here.

**Review by zadjii-msft — APPROVED** [review]
Okay I found an unrelated crash while playing with this, but this seems good to me. Thanks! @cinnamon-msft will come around later to write up a docs page for this

**Comment by cinnamon-msft** [comment]
I could see this in the Panes page of our docs, and also in our Tips and Tricks page. :)

I'm happy to write up the docs for the 1.7 release with this!

File: src/cascadia/TerminalControl/Resources/en-US/Resources.resw
Comment (cinnamon-msft) [inline src/cascadia/TerminalControl/Resources/en-US/Resources.resw:201]:
```suggestion
    <value>Read-only mode is enabled.</value>

```

The tiniest of nits

**Comment by Don-Vito** [comment]
@zadjii-msft - what are the next steps here?

**Comment by Don-Vito** [comment]
> 
> 
> I could see this in the Panes page of our docs, and also in our Tips and Tricks page. :)
> 
> I'm happy to write up the docs for the 1.7 release with this!

I have created an issue in the docs repo: https://github.com/MicrosoftDocs/terminal/issues/241

**Comment by PankajBhojwani** [comment]
> what are the next steps here? I am worried about conflicts with #9024 😊

I anticipate that #9024 is going to take _a while_ to both be finished and get signoffs, so I'll go ahead and say we should merge this first and I'll deal with the conflicts

**Comment by Don-Vito** [comment]
@zadjii-msft - can I somehow help moving this forward?

**Comment by zadjii-msft** [comment]
Whoops, that's my B for not slapping automerge on this. I've resolved the conflict with #9034, and hopefully that will automerge itself in ~15 minutes. Sorry about that!

**Comment by Don-Vito** [comment]
@zadjii-msft - there was some build error, so I re-merged.

**Comment by Don-Vito** [comment]
Seems to be building now 😛

**Review by DHowett — COMMENTED** [review]
Well done. Thanks!

**Comment by rluvaton** [comment]
Great feature! In which version this will be available?

**Comment by DHowett** [comment]
Any feature pull request that is merged will be available in the preview build that comes out after it was merged. Usually. :)

**Comment by rluvaton** [comment]
> Any feature pull request that is merged will be available in the preview build that comes out after it was merged. Usually. :)

Yeah, but I don't see it in the release notes

**Comment by DHowett** [comment]
Preview version 1.7 hasn’t been released yet, so I am not sure what release notes you’re talking about 😄

**Comment by rluvaton** [comment]
@DHowett Ohh, thanks! by the way, this is the BEST terminal I ever use! thank you so much for the active development and bringing cool and useful features in no time!

**Pull request merged**
